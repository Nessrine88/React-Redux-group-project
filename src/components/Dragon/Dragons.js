// Dragons.js

import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  fetchDragons,
  reserveDragon,
  cancelDragonReservation,
  selectDragons,
  selectReservedDragons,
} from '../../redux/dragons/dragonsSlice';
import './Dragon.css';

function Dragons() {
  const dragons = useSelector(selectDragons);
  const reservedDragons = useSelector(selectReservedDragons);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchDragons());
  }, [dispatch]);

  useEffect(() => {
    const reservedDragonIds = JSON.parse(sessionStorage.getItem('reservedDragons') || '[]');
    reservedDragonIds.forEach((id) => dispatch(reserveDragon({ id })));
  }, [dispatch]);

  const handleReserveDragon = (id) => {
    dispatch(reserveDragon({ id }));

    const reservedDragonIds = JSON.parse(sessionStorage.getItem('reservedDragons') || '[]');
    if (!reservedDragonIds.includes(id)) {
      reservedDragonIds.push(id);
      sessionStorage.setItem('reservedDragons', JSON.stringify(reservedDragonIds));
    }
  };

  const handleCancelReservation = (id) => {
    dispatch(cancelDragonReservation({ id }));

    const reservedDragonIds = JSON.parse(sessionStorage.getItem('reservedDragons') || '[]');
    const updatedReservedDragonIds = reservedDragonIds.filter((dragonId) => dragonId !== id);
    sessionStorage.setItem('reservedDragons', JSON.stringify(updatedReservedDragonIds));
  };

  return (
    <div className="dragons-container">
      <h1 className="dragons-header">Dragons</h1>
      <ul className="dragons-list">
        {dragons.map((dragon) => (
          <li className="dragon-item" key={dragon.id}>
            <img className="dragon-image" src={dragon.flickr_images[0]} alt={dragon.name} />
            <div className="dragon-details">
              <h2 className="dragon-name">{dragon.name}</h2>
              <p className="dragon-type">
                Type:
                {' '}
                {dragon.type}
              </p>
              <p className="dragon-description">
                Description:
                {' '}
                {dragon.description}
              </p>
              {reservedDragons[dragon.id] ? (
                <button className="cancel-button" type="button" onClick={() => handleCancelReservation(dragon.id)}>
                  Cancel Reservation
                </button>
              ) : (
                <button className="reserve-button" type="button" onClick={() => handleReserveDragon(dragon.id)}>
                  Reserve
                </button>
              )}
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
}

export default Dragons;
