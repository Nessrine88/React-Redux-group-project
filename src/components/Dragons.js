// Dragons.js
import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  fetchDragons, reserveDragon, cancelDragonReservation, selectDragons,
} from '../redux/dragons/dragonsSlice';

function Dragons() {
  const dragons = useSelector(selectDragons);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchDragons());
  }, [dispatch]);

  const handleReserveDragon = (id) => {
    dispatch(reserveDragon({ id }));
  };

  const handleCancelReservation = (id) => {
    dispatch(cancelDragonReservation({ id }));
  };

  return (
    <div>
      <h2>Dragons</h2>
      <ul>
        {dragons.map((dragon) => (
          <li key={dragon.id}>
            <h3>{dragon.name}</h3>
            <p>{dragon.description}</p>
            {dragon.reserved ? (
              <div>
                <p>Status: Reserved</p>
                <button type="button" onClick={() => handleCancelReservation(dragon.id)}>
                  Cancel Reservation
                </button>
              </div>
            ) : (
              <div>
                <p>Status: Not Reserved</p>
                <button type="button" onClick={() => handleReserveDragon(dragon.id)}>
                  Reserve Dragon
                </button>
              </div>
            )}
          </li>
        ))}
      </ul>
    </div>
  );
}

export default Dragons;
