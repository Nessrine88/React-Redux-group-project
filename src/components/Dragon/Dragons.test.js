import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import Dragons from './Dragons'; // Update the path as needed
import dragonsReducer from '../../redux/dragons/dragonsSlice';

const store = configureStore({
  reducer: {
    dragons: dragonsReducer,
  },
});

test('renders Dragons component', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <Dragons />
    </Provider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
