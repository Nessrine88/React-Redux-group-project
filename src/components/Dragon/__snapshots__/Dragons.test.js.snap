// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Dragons component 1`] = `
<DocumentFragment>
  <div
    class="dragons-container"
  >
    <h1
      class="dragons-header"
    >
      Dragons
    </h1>
    <ul
      class="dragons-list"
    />
  </div>
</DocumentFragment>
`;
